Userland/Libraries/LibWeb/Painting/Paintable.cpp

```diff
@@ -172,4 +172,19 @@ CSSPixelPoint Paintable::box_type_agnostic_position() const
     return position;
 }
 
+Optional<u32> Paintable::get_alpha_mask_id_from_ancestors() const
+{
+    Optional<u32> alpha_mask_id;
+    auto const containing_block = layout_node().containing_block();
+    for (auto* ancestor = containing_block; ancestor; ancestor = ancestor->containing_block()) {
+        if (ancestor->paintable() && ancestor->paintable()->fast_is<PaintableWithLines>()) {
+            auto const& ancestor_with_lines = static_cast<PaintableWithLines const&>(*ancestor->paintable());
+            alpha_mask_id = ancestor_with_lines.get_alpha_mask_id();
+            if (alpha_mask_id.has_value()) break;
+        }
+    }
+
+    return alpha_mask_id;
+}
```

Comment:
traversing the containing block chain is something we try to avoid because:

- it makes painting commands recording noticeably slower and very visible in profiles on any large page
- adds yet another dependency of paintable tree on layout tree

so we are in the process of getting rid of this pattern.
instead, could we assign each paintable that has to be affected by mask corresponding mask id on pre-paint phase, like we do that for overflow clipping and scrolling offsets?
you could check how it works by looking here:
https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibWeb/Painting/ViewportPaintable.cpp#L91
https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibWeb/Painting/ViewportPaintable.cpp#L136

Reply:
This bit of feedback should be addressed now :)